src/file_pipe_log/pipe.rs

```diff
@@ -814,79 +773,50 @@ mod tests {
-
-        // Purge to file 1, this file would be recycled
-        assert_eq!(pipe_log.purge_to(FileId { queue, seq: 2 }).unwrap(), 0);
-        assert_eq!(pipe_log.file_span(queue).0, 2);
-
-        // cannot purge active file
-        assert!(pipe_log.purge_to(FileId { queue, seq: 4 }).is_err());
-
-        // append position
-        let s_content = b"short content".to_vec();
-        let file_handle = pipe_log.append(queue, &s_content).unwrap();
-        pipe_log.maybe_sync(queue, false).unwrap();
-        assert_eq!(file_handle.id.seq, 3);
-        assert_eq!(file_handle.offset, header_size);
-
-        let file_handle = pipe_log.append(queue, &s_content).unwrap();
-        pipe_log.maybe_sync(queue, false).unwrap();
-        assert_eq!(file_handle.id.seq, 3);
-        assert_eq!(
-            file_handle.offset,
-            header_size as u64 + s_content.len() as u64
-        );
-
-        let content_readed = pipe_log
-            .read_bytes(FileBlockHandle {
-                id: FileId { queue, seq: 3 },
-                offset: header_size as u64,
-                len: s_content.len(),
-            })
+        let fs = Arc::new(ObfuscatedFileSystem::default());
+        let pipe_log = new_test_pipe(&cfg, queue, fs.clone()).unwrap();
+        assert_eq!(pipe_log.file_span(), (1, 1));
+
+        let content: Vec<u8> = vec![b'a'; 16];
+        let mut handles = Vec::new();
+        for _ in 0..10 {
+            handles.push(pipe_log.append(&content).unwrap());
+            pipe_log.maybe_sync(true).unwrap();
+        }
+        let (first, last) = pipe_log.file_span();
+        assert_eq!(pipe_log.purge_to(last).unwrap() as u64, last - first);
+        // Try to read stale file.
+        for handle in handles {
+            assert!(pipe_log.read_bytes(handle).is_err());
+            // Bypass pipe log
+            let mut reader = build_file_reader(
+                fs.as_ref(),
+                Arc::new(fs.open(handle.id.build_file_path(path)).unwrap()),
+            )
             .unwrap();
-        assert_eq!(content_readed, s_content);
-        // try to fetch abnormal entry
-        let abnormal_content_readed = pipe_log.read_bytes(FileBlockHandle {
-            id: FileId { queue, seq: 12 }, // abnormal seq
-            offset: header_size as u64,
-            len: s_content.len(),
-        });
-        assert!(abnormal_content_readed.is_err());
-
-        // leave only 1 file to truncate
-        pipe_log.purge_to(FileId { queue, seq: 3 }).unwrap();
-        assert_eq!(pipe_log.file_span(queue), (3, 3));
-
-        // fetch active file
-        let file_context = pipe_log.fetch_active_file(LogQueue::Append);
-        assert_eq!(file_context.version, Version::V2);
-        assert_eq!(file_context.id.seq, 3);
+            assert_eq!(reader.read(handle).unwrap(), content);
+            // Delete file so that it cannot be reused.
+            fs.delete(handle.id.build_file_path(path)).unwrap();
```

Comment:
The `deletion` is redundant here, I think.
The implementation of `reuse` in ObfuscatedFileSystem has already do `delete(src)` and `create(dst)`. And I think this case should testing the following cases:
* Builds `pipe` with `DefaultFileSystem` for recycling, validates that `reuse` is just `rename` a stale file for `old_filepath` to `new_filepath`.
* Builds `pipe` with `ObfuscatedFileSystem` for recycling, validates that `reuse` will delete the old file and create a new file. The new file is empty.

Reply:
It's not redundant. It is to make the next `reuse` fails.

There's no need to test the behavior of a specific file system because it can have arbitrary user implementation.